Github authentication error should not cause 500 error

When using Github authentication, auth errors from Github API causes 500 error for user. We should handle authentication errors and return proper status code in that case 401.

```
artipie         | Caused by: java.lang.AssertionError: HTTP response status is not equal to 200:
artipie         | 401 Unauthorized [https://api.github.com/user]
artipie         | Server: GitHub.com
artipie         | Date: Thu, 30 Jul 2020 14:39:46 GMT
artipie         | Content-Type: application/json; charset=utf-8
artipie         | Content-Length: 83
artipie         | Status: 401 Unauthorized
artipie         | X-GitHub-Media-Type: github.v3
artipie         | X-RateLimit-Limit: 60
artipie         | X-RateLimit-Remaining: 55
artipie         | X-RateLimit-Reset: 1596123034
artipie         | Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset
artipie         | Access-Control-Allow-Origin: *
artipie         | Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
artipie         | X-Frame-Options: deny
artipie         | X-Content-Type-Options: nosniff
artipie         | X-XSS-Protection: 1; mode=block
artipie         | Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
artipie         | Content-Security-Policy: default-src 'none'
artipie         | Vary: Accept-Encoding, Accept, X-Requested-With
artipie         | 
artipie         | {"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}
artipie         | Expected: HTTP response with status 200
artipie         |      but: was <401 Unauthorized [https://api.github.com/user]
artipie         | Server: GitHub.com
artipie         | Date: Thu, 30 Jul 2020 14:39:46 GMT
artipie         | Content-Type: application/json; charset=utf-8
artipie         | Content-Length: 83
artipie         | Status: 401 Unauthorized
artipie         | X-GitHub-Media-Type: github.v3
artipie         | X-RateLimit-Limit: 60
artipie         | X-RateLimit-Remaining: 55
artipie         | X-RateLimit-Reset: 1596123034
artipie         | Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset
artipie         | Access-Control-Allow-Origin: *
artipie         | Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
artipie         | X-Frame-Options: deny
artipie         | X-Content-Type-Options: nosniff
artipie         | X-XSS-Protection: 1; mode=block
artipie         | Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
artipie         | Content-Security-Policy: default-src 'none'
artipie         | Vary: Accept-Encoding, Accept, X-Requested-With
artipie         | 
artipie         | {"message":"Bad credentials","documentation_url":"https://developer.github.com/v3"}>
artipie         | 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
artipie         | 	at com.jcabi.http.response.RestResponse.assertStatus(RestResponse.java:111)
artipie         | 	at com.jcabi.github.RtJson.fetch(RtJson.java:77)
artipie         | 	at com.jcabi.github.RtUser.json(RtUser.java:166)
artipie         | 	at com.jcabi.github.RtUser.login(RtUser.java:121)
artipie         | 	at com.artipie.auth.GithubAuth.lambda$new$0(GithubAuth.java:65)
artipie         | 	at com.artipie.auth.GithubAuth.user(GithubAuth.java:86)
artipie         | 	at com.artipie.auth.ChainedAuth.user(ChainedAuth.java:74)
artipie         | 	at com.artipie.auth.LoggingAuth.user(LoggingAuth.java:68)
artipie         | 	at com.artipie.http.auth.BasicIdentities.lambda$user$3(BasicIdentities.java:67)
artipie         | 	at java.base/java.util.Optional.flatMap(Unknown Source)
artipie         | 	at com.artipie.http.auth.BasicIdentities.user(BasicIdentities.java:67)
artipie         | 	at com.artipie.http.auth.SliceAuth.response(SliceAuth.java:85)
artipie         | 	at com.artipie.http.rt.RtRulePath.response(RtRulePath.java:72)
artipie         | 	at com.artipie.http.rt.SliceRoute.lambda$response$0(SliceRoute.java:86)
artipie         | 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
artipie         | 	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Unknown Source)
artipie         | 	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(Unknown Source)
artipie         | 	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(Unknown Source)
artipie         | 	at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
artipie         | 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
artipie         | 	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(Unknown Source)
artipie         | 	at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
artipie         | 	at java.base/java.util.stream.ReferencePipeline.findFirst(Unknown Source)
artipie         | 	at com.artipie.http.rt.SliceRoute.response(SliceRoute.java:89)
artipie         | 	at com.artipie.http.Slice$Wrap.response(Slice.java:81)
artipie         | 	at com.artipie.http.slice.TrimPathSlice.response(TrimPathSlice.java:112)
artipie         | 	at com.artipie.http.async.AsyncSlice.lambda$response$0(AsyncSlice.java:63)
artipie         | 	... 15 more
```

--- a/src/main/java/com/artipie/auth/GithubAuth.java
+++ b/src/main/java/com/artipie/auth/GithubAuth.java
@@ -13,6 +13,9 @@ import java.util.Optional;
 import java.util.function.Function;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import org.cactoos.scalar.Not;
+import org.cactoos.scalar.Unchecked;
+import org.cactoos.text.Contains;
 
 /**
  * GitHub authentication uses username prefixed by provider name {@code github.com}
@@ -64,11 +67,23 @@ public final class GithubAuth implements Authentication {
         Optional<Authentication.User> result = Optional.empty();
         final Matcher matcher = GithubAuth.PTN_NAME.matcher(username);
         if (matcher.matches()) {
-            final String login = this.github.apply(password).toLowerCase(Locale.US);
-            if (
-                Objects.equals(login, matcher.group(1).toLowerCase(Locale.US))
-            ) {
-                result = Optional.of(new Authentication.User(login));
+            try {
+                final String login = this.github.apply(password).toLowerCase(Locale.US);
+                if (
+                    Objects.equals(login, matcher.group(1).toLowerCase(Locale.US))
+                ) {
+                    result = Optional.of(new Authentication.User(login));
+                }
+            } catch (final AssertionError error) {
+                if (
+                    new Unchecked<>(
+                        new Not(
+                            new Contains(error.getMessage(), "401 Unauthorized")
+                        )
+                    ).value()
+                ) {
+                    throw error;
+                }
             }
         }
         return result;

--- a/src/main/java/com/artipie/auth/GithubAuth.java
+++ b/src/main/java/com/artipie/auth/GithubAuth.java
@@ -4,6 +4,7 @@
  */
 package com.artipie.auth;
 
+import com.artipie.ArtipieException;
 import com.artipie.http.auth.Authentication;
 import com.jcabi.github.RtGithub;
 import java.io.IOException;
@@ -13,9 +14,6 @@ import java.util.Optional;
 import java.util.function.Function;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
-import org.cactoos.scalar.Not;
-import org.cactoos.scalar.Unchecked;
-import org.cactoos.text.Contains;
 
 /**
  * GitHub authentication uses username prefixed by provider name {@code github.com}
@@ -75,14 +73,9 @@ public final class GithubAuth implements Authentication {
                     result = Optional.of(new Authentication.User(login));
                 }
             } catch (final AssertionError error) {
-                if (
-                    new Unchecked<>(
-                        new Not(
-                            new Contains(error.getMessage(), "401 Unauthorized")
-                        )
-                    ).value()
-                ) {
-                    throw error;
+                if (error.getMessage() == null
+                    || !error.getMessage().contains("401 Unauthorized")) {
+                    throw new ArtipieException(error);
                 }
             }
         }

--- a/src/test/java/com/artipie/auth/GithubAuthTest.java
+++ b/src/test/java/com/artipie/auth/GithubAuthTest.java
@@ -4,6 +4,7 @@
  */
 package com.artipie.auth;
 
+import com.artipie.ArtipieException;
 import com.artipie.http.auth.Authentication;
 import java.util.Optional;
 import org.cactoos.text.Joined;
@@ -56,7 +57,7 @@ final class GithubAuthTest {
     @Test
     void shouldThrownExceptionWhenAssertionErrorIsHappened() {
         Assertions.assertThrows(
-            AssertionError.class,
+            ArtipieException.class,
             () -> new GithubAuth(
                 token -> {
                     throw new AssertionError("Any error");

--- a/src/test/java/com/artipie/auth/GithubAuthTest.java
+++ b/src/test/java/com/artipie/auth/GithubAuthTest.java
@@ -5,8 +5,11 @@
 package com.artipie.auth;
 
 import com.artipie.http.auth.Authentication;
+import java.util.Optional;
+import org.cactoos.text.Joined;
 import org.hamcrest.MatcherAssert;
 import org.hamcrest.core.IsEqual;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 /**
@@ -30,6 +33,35 @@ final class GithubAuthTest {
                 }
             ).user("github.com/UsEr", secret).orElseThrow(),
             new IsEqual<>(new Authentication.User("user"))
+        );
+    }
+
+    @Test
+    void shouldReturnOptionalEmptyWhenRequestIsUnauthorized() {
+        MatcherAssert.assertThat(
+            new GithubAuth(
+                token -> {
+                    throw new AssertionError(
+                        new Joined(
+                            "HTTP response status is not equal to 200:\n",
+                            "401 Unauthorized [https://api.github.com/user]"
+                        )
+                    );
+                }
+            ).user("github.com/bad_user", "bad_secret"),
+            new IsEqual<>(Optional.empty())
+        );
+    }
+
+    @Test
+    void shouldThrownExceptionWhenAssertionErrorIsHappened() {
+        Assertions.assertThrows(
+            AssertionError.class,
+            () -> new GithubAuth(
+                token -> {
+                    throw new AssertionError("Any error");
+                }
+            ).user("github.com/user", "pwd")
         );
     }
 }

